pkg/controllers/work/apply_controller.go

```diff
@@ -383,13 +399,50 @@ func (r *ApplyWorkReconciler) applyUnstructured(ctx context.Context, gvr schema.
 	}
 
 	// We only try to update the object if its spec hash value has changed.
-	if manifestObj.GetAnnotations()[manifestHashAnnotation] != curObj.GetAnnotations()[manifestHashAnnotation] {
+	if manifestObj.GetAnnotations()[ManifestHashAnnotation] != curObj.GetAnnotations()[ManifestHashAnnotation] {
+		// we need to merge the owner reference between the current and the manifest since we support one manifest
+		// belong to multiple work so it contains the union of all the appliedWork.
+		manifestObj.SetOwnerReferences(mergeOwnerReference(curObj.GetOwnerReferences(), manifestObj.GetOwnerReferences()))
+		// record the raw manifest with the hash annotation in the manifest.
+		if err = setModifiedConfigurationAnnotation(manifestObj); err != nil {
+			return nil, ManifestNoChangeAction, err
+		}
+		annotations := manifestObj.GetAnnotations()
+		klog.V(2).InfoS("validating annotation size for manifest",
+			"gvr", gvr, "manifest", manifestRef)
+		if err = validation.ValidateAnnotationsSize(annotations); err != nil {
+			klog.ErrorS(err, "not using three way merge for manifest removing last applied config annotation",
+				"gvr", gvr, "obj", manifestRef)
+			annotations[LastAppliedConfigAnnotation] = ""
+			manifestObj.SetAnnotations(annotations)
```

Comment:
As i mentioned before, this part could be part of the setModifiedConfigurationAnnotation func. then you don't need to duplicate the code in two places.

For the update, you only need to check if the LastAppliedConfigAnnotation is empty, server side apply, otherwise, three way merge.